### ISSUE TITLE
Multiserver Support (and v0.9.0 Bug Catching Spree)

### ISSUE THREAD

**Issue by ntimo** [SimonOrJ/CoreProtect-Lookup-Web-Interface#2]
Hi there,

it would be really awesome if I could add Multiple Server with one panel. So I can specify 2 database connections in the config and name the servers and then select the server to search trough that would be so so amazing.

Thx
Timo

**Comment by chuushi** [comment]
I was thinking about that, but I haven't acted upon it yet so far.  It shouldn't be too hard to do.  I would recommend having some degree of per-server configuration system.

I'm currently rewriting parts of the code right now to make it more understandable by writing in more comments, using more standard functions, updating some packages used in the project, etc. which should be up within the next 4 days.  If you want to give it a shot at adding it, you can try it after I commit the update.

**Comment by ntimo** [comment]
Hi,
I would try to code it myself if I could :) I hope you will add support for this soon ;) 

Thx
Timo

**Comment by chuushi** [comment]
Ah, gotcha.  I'll be on it soon!

**Comment by ntimo** [comment]
Nice, thank you so much. For your amazing panel

**Comment by chuushi** [comment]
Anytime :)

Thank you for taking your time to make an issue! :D

**Comment by ntimo** [comment]
Sure, I will make another one :D

**Comment by ntimo** [comment]
@SimonOrJ do you have any news about your v0.9.0 milestone ;)

**Comment by chuushi** [comment]
I've been busy with life chaos... Sorry about the delay! ._.

I'll make sure to get it done before Saturday. I'm so sorry about that.

**Comment by chuushi** [comment]
As for the update, I've been replacing the variables in `conn.php` so far.  I plan on fixing the `index.php` URL updating and an option to do an in-browser (re)configuration before I release v0.9.0. :)

**Comment by chuushi** [comment]
I've vastly underestimated the time I can take working on this project and the amount of work to do to create a new configuration system.  It's getting pretty close to becoming done, though.

**Comment by ntimo** [comment]
@SimonOrJ what exactly do you mean with its getting pretty close? Can't wait to test it out. This would be an really awesome feature!

**Comment by chuushi** [comment]
Sorry, I lied by saying "pretty close".  I'm sorry to keep you waiting by pushing the Milestone deadline three times!  The version v0.9.0 is definitely going to be done by Wednesday for sure.

This will use a different configuration scheme, and the best part is that this can be configured by editing the files or through the web interface!

**Comment by ntimo** [comment]
@SimonOrJ Oh okay, that sounds like a nice new config system. But will it be able to handle more then one server with one interface? I am looking forward to Wednesday.

**Comment by ntimo** [comment]
@SimonOrJ Hi, I just tried out the new version from the pre branch. And I always get an error when trying to add an server. And there is no option to select the server on the search page. And maybe the setup page should only be accessible after logging in.

But I also wanted to thank you, for your awesome support.

**Comment by chuushi** [comment]
The server addition error was reported because the server adding part wasn't fully operational.  The software didn't use the new configuration scheme at that time, but it does now!

The login is disabled system wide for now, though.
6764fb0359a053e7c827fe2936459ec35ef48dcf

Edit: Figured out the whole login thing. I made it mine now, and it works well. Mwahaha!

**Comment by chuushi** [comment]
The multi-server read/cache support is basically complete! :D

Now all that needs to be done is deleting cache by server button and upgrading the JS support. The dynmap support is busted as it stands now...

**Comment by ntimo** [comment]
@SimonOrJ when I try to login and then open the setup.php I get: Warning: 

```
require(/res/php/webtemplate.php): failed to open stream: No such file or directory in /var/www/vhosts/xxxxxxxx/xxxxxxxxx/setup.php on line 253

Fatal error: require(): Failed opening required '/res/php/webtemplate.php' (include_path='.:/opt/plesk/php/7.0/share/pear') in /var/www/vhosts/xxxxxxxx/xxxxxxxxx/setup.php on line 253
```

And what is this BanManger? :O Thats in the top Menu bar.

**Comment by ntimo** [comment]
@SimonOrJ after manually adding the server config files. I get an error when trying to search for something in the database 

`E  Unexpected Error 7: The lookup script was unable to send a proper response.`

**Comment by chuushi** [comment]
Oh, that's a test navigation.  Banmanager's a separate Minecraft plugin.

I'll look into the database error.

**Comment by ntimo** [comment]
@SimonOrJ Hi, now the  setup.php works nice thank you :) I hope that the lookup will work soon too :)

Also nice work ;)

**Comment by chuushi** [comment]
I think I found the problem to the error.  Does it work now?

Also, thank you for supporting the project! :D

**Comment by ntimo** [comment]
@SimonOrJ now I get this on the setup.php when trying to add a new server:

`Fatal error: Uncaught Error: Class 'ChromePhp' not found in /var/www/vhosts/xxxxxx/xxxxxxx/setup.php:95 Stack trace: #0 {main} thrown in /var/www/vhosts/xxxxxx/xxxxxxx/setup.php on line 95`

**Comment by ntimo** [comment]
@SimonOrJ okay when using the load more button I get:

`E  Unexpected Error 5: Server variable does not exist.`

**Comment by ntimo** [comment]
@SimonOrJ also the username and blocks prediction does not work :(

**Comment by ntimo** [comment]
@SimonOrJ also when searching for a block like minecraft:dirt I get the error:

`E  Unexpected Error 7: The lookup script was unable to send a proper response.`

**Comment by chuushi** [comment]
The "Load More" doesn't work yet; it's still being worked on.

~~The username and block predictions will only work for the blocks that were searched before.~~
EDIT: Actually, this code needed some updating.

I've just fixed the first two problems. I'll commit it now. :)

**Comment by ntimo** [comment]
Now when I try to search (blocks and username) it says: Server is not defined.

The setup page works now but does not redirects you after adding a server.

**Comment by chuushi** [comment]
Hmm, yeah, more work needs to be done.  I'm working on the JavaScript to fix the server undefined issue right now.

The setup page is in the lower priority since it can be configured directly from the source files.

**Comment by ntimo** [comment]
Hi, do you have any news for the JavaScript fix?

**Comment by ntimo** [comment]
@SimonOrJ Now I get please submit a lookup when trying to look something up :( And this error ate the end of the page

`Fatal error: Uncaught Error: Call to undefined method Login::username() in /var/www/vhosts/xxxxxx/xxxxxxx/index.php:271 Stack trace: #0 {main} thrown in /var/www/vhosts/xxxxxx/xxxxxxx/index.php on line 271`

**Comment by chuushi** [comment]
I finally got some sort of a fix complete and committed.

**Comment by ntimo** [comment]
Hi, I still get the error: E    Unexpected Error 7: The lookup script was unable to send a proper response. When I try to look something up.

**Comment by chuushi** [comment]
Huh, that's weird.  What do you get when you go into `web/conn.php` from the browser?  It should be in this format:

``` json
[
    {"status": "An Object about server status"},
    ["An array containing results in Object format"]
]
```

If there's an error, the array of results doesn't get returned, or the top of the document will have `<br/>` and some PHP error report.

You can redact the "array containing results" if it works successfully.

**Comment by ntimo** [comment]
I get this:
`[{"status":5,"reason":"Server variable does not exist.","duration":0.00054001808166504}]`

**Comment by chuushi** [comment]
Hmm... Try putting `?server=(whatever your server name is)` at the end of the URL.

**Comment by ntimo** [comment]
Can I send you the login information and you can try it out?

**Comment by ntimo** [comment]
I send you a e-mail to the adress that was in your github profile with the url and login information.

**Comment by chuushi** [comment]
Ah, you have to redo the configuration.  There were some changes in the configuration in this commit: b3241c5d9b95238b4b037d7840e4742d69124594

For an easy solution, change the word "form" to "flag".

**Comment by ntimo** [comment]
Okay, where exactly do I have to change this?

**Comment by ntimo** [comment]
Okay, I found it and now it works 👍

**Comment by ntimo** [comment]
It seems like I can't search for already Rolled back data

**Comment by chuushi** [comment]
It should be fixed now.  It was an overlooked bug.

**Comment by ntimo** [comment]
Now I get the: `E   Unexpected Error 7: The lookup script was unable to send a proper response.` Error

**Comment by ntimo** [comment]
okay, now the lookup works but the search for rolled back blocks does not work :(

**Comment by ntimo** [comment]
The load more button also does not work and just give out an `E`

**Comment by chuushi** [comment]
Yeah, it's a caveat.

I'll look into the rollback issue.  I wasn't around my computer since 10 hours ago, so I couldn't get to anything yet.

By the way, I like to keep my developmental changelog on commit messages, so it'll be good to check out https://github.com/SimonOrJ/CoreProtect-Lookup-Web-Interface/commits/0.9.0-preview whwnever you download or `git pull` the updated files.

**Comment by chuushi** [comment]
Okay, the mentioned errors sans the Load More caveat should be fixed now!

**Comment by ntimo** [comment]
Hi there,

the load more button this does not work :( It just displays an `E`

**Comment by chuushi** [comment]
I've made multiple fixes to the form.  Everything should work pretty well now!

Now I'm going to focus on making JS for the setup.php forms.

**Comment by ntimo** [comment]
Now everything works thank you :)

**Comment by chuushi** [comment]
Phew, that's awesome to hear :D

I've updated the project to make use of the setup.php page.  There's some documentation work to be done there, but it's usable for now.

**Comment by chuushi** [comment]
Feel free to open up another pull request if anything else goes wrong!  I'll go ahead and pull the changes to the main branch after I add some documentation soon!

Thank you so much for working with me throughout so far. :)
